typescript/token/src/adapters/EvmTokenAdapter.ts

```diff
@@ -0,0 +1,213 @@
+import { BigNumber, PopulatedTransaction } from 'ethers';
+
+import {
+  BaseEvmAdapter,
+  ChainName,
+  MultiProtocolProvider,
+} from '@hyperlane-xyz/sdk';
+import {
+  Address,
+  Domain,
+  addressToByteHexString,
+  addressToBytes32,
+  bytes32ToAddress,
+  isValidAddressEvm,
+  strip0x,
+} from '@hyperlane-xyz/utils';
+
+import { MinimalTokenMetadata } from '../config';
+import {
+  ERC20,
+  ERC20__factory,
+  HypERC20,
+  HypERC20Collateral__factory,
+  HypERC20__factory,
+} from '../types';
+
+import {
+  IHypTokenAdapter,
+  ITokenAdapter,
+  TransferParams,
+  TransferRemoteParams,
+} from './ITokenAdapter';
+
+// Interacts with native currencies
+export class EvmNativeTokenAdapter
+  extends BaseEvmAdapter
+  implements ITokenAdapter
+{
+  async getBalance(address: Address): Promise<string> {
+    if (!isValidAddressEvm(address)) return '0';
+    const balance = await this.getProvider().getBalance(address);
+    return balance.toString();
+  }
+
+  async getMetadata(): Promise<MinimalTokenMetadata> {
+    // TODO get metadata from chainMetadata config
+    throw new Error('Metadata not available to native tokens');
+  }
+
+  async populateApproveTx(
+    _params: TransferParams,
+  ): Promise<PopulatedTransaction> {
+    throw new Error('Approve not required for native tokens');
+  }
+
+  async populateTransferTx({
+    weiAmountOrId,
+    recipient,
+  }: TransferParams): Promise<PopulatedTransaction> {
+    const value = BigNumber.from(weiAmountOrId);
+    return { value, to: recipient };
+  }
+}
+
+// Interacts with ERC20/721 contracts
+export class EvmTokenAdapter<T extends ERC20 = ERC20>
+  extends EvmNativeTokenAdapter
```

Comment:
token extends native?

Reply:
would kinda expect the inverse

Reply:
Conceptually sure, this is just what was convenient implementation-wise. I'm open to refactoring if this rubs you the wrong way